File: ruby/dhondt.rb
class Party
    attr_accessor :name, :votes, :seats
  
    def initialize(name, votes, seats = 0)
      @name = name
      @votes = votes
      @seats = seats
    end
  end
  
  class Quotient
    attr_accessor :party_index, :value
  
    def initialize(party_index, value)
      @party_index = party_index
      @value = value
    end
  end
  
  def get_random_int(min, max)
    rand(min..max)
  end
  
  def get_random_name(length)
    characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
    name = ''
    length.times do
      name += characters[rand(characters.length)]
    end
    name
  end
  
  def generate_random_parties(num_parties, max_votes)
    parties = []
    num_parties.times do
      name_length = get_random_int(1, 10)
      party_name = get_random_name(name_length)
      votes = get_random_int(0, max_votes)
      parties << Party.new(party_name, votes)
    end
    parties
  end
  
  def d_hondt(parties, total_seats)
    # Initialize seats for each party to 0
    parties.each do |party|
      party.seats = 0
    end
  
    quotients = []
  
    # Generate all quotients
    parties.each_with_index do |party, i|
      (1..total_seats).each do |j|
        quotients << Quotient.new(i, party.votes.to_f / j)
      end
    end
  
    # Sort quotients in descending order
    quotients.sort_by! { |q| -q.value }
  
    # Allocate seats based on the highest quotients
    total_seats.times do |i|
      party_index = quotients[i].party_index
      parties[party_index].seats += 1
    end
  
    parties
  end
  
  num_parties = 5
  max_votes = 1000
  parties = generate_random_parties(num_parties, max_votes)
  
  total_seats = 10
  result = d_hondt(parties, total_seats)
  puts result.map { |party| "#{party.name}: #{party.votes} votes, #{party.seats} seats" }
  

File: swift/dhondt.swift
import Foundation

struct Party {
    var name: String
    var votes: Int
    var seats: Int
}

struct Quotient {
    var partyIndex: Int
    var value: Double
}

func getRandomInt(min: Int, max: Int) -> Int {
    return Int.random(in: min...max)
}

func getRandomName(length: Int) -> String {
    let characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
    return String((0..<length).map { _ in characters.randomElement()! })
}

func generateRandomParties(numParties: Int, maxVotes: Int) -> [Party] {
    var parties: [Party] = []
    for _ in 0..<numParties {
        let nameLength = getRandomInt(min: 1, max: 10)
        let partyName = getRandomName(length: nameLength)
        let votes = getRandomInt(min: 0, max: maxVotes)
        parties.append(Party(name: partyName, votes: votes, seats: 0))
    }
    return parties
}

func dHondt(parties: [Party], totalSeats: Int) -> [Party] {
    var parties = parties
    
    // Initialize seats for each party to 0
    for i in 0..<parties.count {
        parties[i].seats = 0
    }
    
    var quotients: [Quotient] = []
    
    // Generate all quotients
    for i in 0..<parties.count {
        for j in 1...totalSeats {
            quotients.append(Quotient(partyIndex: i, value: Double(parties[i].votes) / Double(j)))
        }
    }
    
    // Sort quotients in descending order
    quotients.sort(by: { $0.value > $1.value })
    
    // Allocate seats based on the highest quotients
    for i in 0..<totalSeats {
        let partyIndex = quotients[i].partyIndex
        parties[partyIndex].seats += 1
    }
    
    return parties
}

let numParties = 5
let maxVotes = 1000
let parties = generateRandomParties(numParties: numParties, maxVotes: maxVotes)

let totalSeats = 10
let result = dHondt(parties: parties, totalSeats: totalSeats)
print(result)
